From a2aef0fef652ea2c816535213c02184bd5816283 Mon Sep 17 00:00:00 2001
From: skyjet18 <57456827+skyjet18@users.noreply.github.com>
Date: Sun, 19 Jan 2025 18:42:09 +0100
Subject: [PATCH 19/33] dashdec: improved handling manifests with one segment

- fixed seeking in manifests with one segment per adaptations set
- fixed seeking behing last subtitle (when fake_last_subtitle is set)
- preparation for seeking in live mode (disabled at the moment)
---
 libavformat/dashdec.c | 42 ++++++++++++++++++++++++++++++++++++++----
 1 file changed, 38 insertions(+), 4 deletions(-)

diff --git a/libavformat/dashdec.c b/libavformat/dashdec.c
index c166d3367b..7db298abe1 100644
--- a/libavformat/dashdec.c
+++ b/libavformat/dashdec.c
@@ -36,6 +36,7 @@
 struct fragment {
     int64_t url_offset;
     int64_t size;
+    int seekable;
     char *url;
 };
 
@@ -120,6 +121,7 @@ struct representation {
 
     char *cenc_decryption_key;
     char *cenc_decryption_keys;
+    enum AVMediaType type;
 };
 
 typedef struct DASHContext {
@@ -166,6 +168,7 @@ typedef struct DASHContext {
     int is_init_section_common_subtitle;
 
     uint64_t playlist_load_time;
+    int fake_last_subtitle;
 } DASHContext;
 
 static int ishttp(char *url)
@@ -888,6 +891,7 @@ static int parse_manifest_representation(AVFormatContext *s, const char *url,
     }
 
     rep->parent = s;
+    rep->type = type;
     representation_segmenttemplate_node = find_child_node_by_name(representation_node, "SegmentTemplate");
     representation_baseurl_node = find_child_node_by_name(representation_node, "BaseURL");
     representation_segmentlist_node = find_child_node_by_name(representation_node, "SegmentList");
@@ -1046,6 +1050,7 @@ static int parse_manifest_representation(AVFormatContext *s, const char *url,
         if (!seg->url)
             goto enomem;
         seg->size = -1;
+        seg->seekable = 1;
     } else if (representation_segmentlist_node) {
         // TODO: https://www.brendanlong.com/the-structure-of-an-mpeg-dash-mpd.html
         // http://www-itec.uni-klu.ac.at/dash/ddash/mpdGenerator.php?fragmentlength=15&type=full
@@ -1679,9 +1684,11 @@ try_again:
             }
             seg->size = seg_ptr->size;
             seg->url_offset = seg_ptr->url_offset;
+            seg->seekable = seg_ptr->seekable;
             return seg;
         } else if (c->is_live) {
             refresh_manifest(pls->parent);
+            pls->parent->duration = (int64_t) c->time_shift_buffer_depth * AV_TIME_BASE;
         } else {
             break;
         }
@@ -1692,6 +1699,7 @@ try_again:
 
         if (pls->timelines || pls->fragments) {
             refresh_manifest(pls->parent);
+            pls->parent->duration = (int64_t) c->time_shift_buffer_depth * AV_TIME_BASE;
         }
         if (pls->cur_seq_no <= min_seq_no) {
             av_log(pls->parent, AV_LOG_VERBOSE, "old fragment: cur[%"PRId64"] min[%"PRId64"] max[%"PRId64"]\n", (int64_t)pls->cur_seq_no, min_seq_no, max_seq_no);
@@ -1781,6 +1789,15 @@ static int open_input(DASHContext *c, struct representation *pls, struct fragmen
         av_dict_set_int(&opts, "end_offset", seg->url_offset + seg->size, 0);
     }
 
+    if(seg->seekable && pls->type != AVMEDIA_TYPE_SUBTITLE)
+    {
+        av_dict_set(&opts, "seekable", "1", 0);
+    }
+    else
+    {
+        av_dict_set(&opts, "seekable", "0", 0);
+    }
+
     ff_make_absolute_url(url, c->max_url_size, c->base_url, seg->url);
     av_log(pls->parent, AV_LOG_VERBOSE, "DASH request for url '%s', offset %"PRId64"\n",
            url, seg->url_offset);
@@ -1962,8 +1979,16 @@ static int reopen_demux_for_component(AVFormatContext *s, struct representation
         goto fail;
     }
     ffio_init_context(&pls->pb, avio_ctx_buffer, INITIAL_BUFFER_SIZE, 0,
-                      pls, read_data, NULL, c->is_live ? NULL : seek_data);
-    pls->pb.pub.seekable = 0;
+                      pls, read_data, NULL, seek_data);
+
+    if(pls->type == AVMEDIA_TYPE_SUBTITLE || pls->n_timelines || pls->n_fragments > 1)
+    {
+        pls->pb.pub.seekable = 0;
+    }
+    else
+    {
+        pls->pb.pub.seekable = AVIO_SEEKABLE_NORMAL;
+    }
 
     if ((ret = ff_copy_whiteblacklists(pls->ctx, s)) < 0)
         goto fail;
@@ -1993,6 +2018,11 @@ static int reopen_demux_for_component(AVFormatContext *s, struct representation
     else if (c->cenc_decryption_keys)
         av_dict_set(&in_fmt_opts, "decryption_keys", c->cenc_decryption_keys, 0);
 
+    if(pls->type == AVMEDIA_TYPE_SUBTITLE && c->fake_last_subtitle)
+    {
+        av_dict_set(&in_fmt_opts, "fake_last_subtitle", "1", 0);
+    }
+
     // provide additional information from mpd if available
     ret = avformat_open_input(&pls->ctx, "", in_fmt, &in_fmt_opts); //pls->init_section->url
     av_dict_free(&in_fmt_opts);
@@ -2121,8 +2151,6 @@ static int dash_read_header(AVFormatContext *s)
      * stream. */
     if (!c->is_live) {
         s->duration = (int64_t) c->media_presentation_duration * AV_TIME_BASE;
-    } else {
-        av_dict_set(&c->avio_opts, "seekable", "0", 0);
     }
 
     if(c->n_videos)
@@ -2390,6 +2418,11 @@ static int dash_read_seek(AVFormatContext *s, int stream_index, int64_t timestam
     if ((flags & AVSEEK_FLAG_BYTE) || c->is_live)
         return AVERROR(ENOSYS);
 
+    if (c->is_live && c->time_shift_buffer_depth <=0)
+    {
+        return AVERROR(ENOSYS);
+    }
+
     /* Seek in discarded streams with dry_run=1 to avoid reopening them */
     for (i = 0; i < c->n_videos; i++) {
         if (!ret)
@@ -2437,6 +2470,7 @@ static const AVOption dash_options[] = {
     { "cenc_decryption_keys", "Media decryption keys by KID (hex)", OFFSET(cenc_decryption_keys), AV_OPT_TYPE_STRING, {.str = NULL}, INT_MIN, INT_MAX, .flags = FLAGS },
     { "cenc_decryption_video_key", "Media decryption video key scheme_cenc (hex)", OFFSET(cenc_decryption_video_key), AV_OPT_TYPE_STRING, {.str = NULL}, INT_MIN, INT_MAX, .flags = FLAGS },
     { "cenc_decryption_audio_key", "Media decryption audio key scheme_cenc (hex)", OFFSET(cenc_decryption_audio_key), AV_OPT_TYPE_STRING, {.str = NULL}, INT_MIN, INT_MAX, .flags = FLAGS },
+    { "fake_last_subtitle", "Insert fake last subtitle to prevent subtitle stream end befor audio/video", OFFSET(fake_last_subtitle), AV_OPT_TYPE_INT, { .i64 = 0 }, 0, 1, .flags = FLAGS },
     {NULL}
 };
 
-- 
2.45.2.windows.1

